addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,607 @@
+	}
+	for _, apiResource := range apiResourceList {
+		gv, err := schema.ParseGroupVersion(apiResource.GroupVersion)
+		if err != nil {
+			return nil, err
+		}
+		if gv.Group == gk.Group {
+			for _, res := range apiResource.APIResources { //nolint:gocritic
+				if res.Kind == gk.Kind {
+					r.providerGVR[gk] = &schema.GroupVersionResource{Group: gv.Group, Resource: res.Name, Version: gv.Version}
+					return r.providerGVR[gk], nil
+				}
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find server preferred resource %s/%s", gk.Group, gk.Kind)
+}
+
+// periodicGVRCachesClean invalidates caches used for GVR lookup
+func (r *tanzuClusterBootstrapReconciler) periodicGVRCachesClean() {
+	ticker := time.NewTicker(constants.DiscoveryCacheInvalidateInterval)
+	for {
+		select {
+		case <-r.context.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			r.cachedDiscoveryClient.Invalidate()
+			r.providerGVR = make(map[schema.GroupKind]*schema.GroupVersionResource)
+		}
+	}
+}
+
+// updateValuesFromSecret updates secretRef in valuesFrom
+func (r *tanzuClusterBootstrapReconciler) updateValuesFromSecret(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+	valuesFrom *runtanzuv1alpha3.ValuesFrom, templateNS string, packageShortName string, log logr.Logger) (*corev1.Secret, error) {
+
+	var newSecret *corev1.Secret
+	if valuesFrom.SecretRef != "" {
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{Namespace: templateNS, Name: valuesFrom.SecretRef}
+		if err := r.Get(r.context, key, secret); err != nil {
+			log.Error(err, "unable to fetch secret", "objectkey", key)
+			return nil, err
+		}
+		newSecret = secret.DeepCopy()
+		newSecret.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: clusterapiv1beta1.GroupVersion.String(),
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			},
+		}
+		newSecret.Name = fmt.Sprintf("%s-%s-package", cluster.Name, packageShortName)
+		newSecret.Namespace = bootstrap.Namespace
+		if err := r.Create(r.context, newSecret); err != nil {
+			return nil, err
+		}
+		valuesFrom.SecretRef = newSecret.Name
+	}
+
+	return newSecret, nil
+}
+
+// updateValuesFromProvider updates providerRef in valuesFrom
+func (r *tanzuClusterBootstrapReconciler) updateValuesFromProvider(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+	valuesFrom *runtanzuv1alpha3.ValuesFrom, templateNS string, packageShortName string, log logr.Logger) (*unstructured.Unstructured, error) {
+
+	var newProvider *unstructured.Unstructured
+	if valuesFrom.ProviderRef != nil {
+		gvr, err := r.getGVR(schema.GroupKind{Group: *valuesFrom.ProviderRef.APIGroup, Kind: valuesFrom.ProviderRef.Kind})
+		if err != nil {
+			log.Error(err, "failed to getGVR")
+			return nil, err
+		}
+		provider, err := r.dynamicClient.Resource(*gvr).Namespace(templateNS).Get(r.context, valuesFrom.ProviderRef.Name, metav1.GetOptions{})
```

Comment:
question: remind me again how these initial `providerRef`'s get onto the cluster? is this with the initial tkg-management package install?

Reply:
The CR objects should be part of the TKR package.

Reply:
cool. so like they'd be a "template" `PinnipedConfig` from which the rest of the `PinnipedConfig` CRs would get created (correct me if i am wrong)

is there guidance on whether we should prevent a controller from reconciling this "template" `PinnipedConfig` CR into a child `Secret`?

this is what i am imagining happening:
1. tkg-management package installed (including `PinnipedConfig` "template" CR and pinniped-config-controller)
2. pinniped-config-controller starts running and sees `PinnipedConfig` "template" CR
3. pinniped-config-controller creates addon `Secret` with data values from `PinnipedConfig` "template" CR
4. this extra addon `Secret` is never used by anything on the cluster

Reply:
The template object is never associated to a cluster(via ownerReferences or naming convention). If you look at https://github.com/vmware-tanzu/tanzu-framework/pull/1555 the secretRef should be set on status only if there is an  associated cluster object.